Hierarchies-react: add localization support (#583)
iTwin/presentation: packages/hierarchies-react/README.md

  return <TreeRenderer rootNodes={rootNodes} {...state} />;
}
```

## Localization

> Localization can be enabled for `TreeRenderer` component and `useTree` and `useUnifiedSelectionTree` hooks by providing an object with localized strings that will be used instead of the default english ones.

Reviewer: ```suggestion
Localization can be enabled for `TreeRenderer` component and `useTree` and `useUnifiedSelectionTree` hooks by providing an object with localized strings that will be used instead of the default English ones.
```

Example:

```tsx
const localizedStrings = {
  loading: "Loading...",
  filterHierarchyLevel: "Apply filter",
  clearHierarchyLevelFilter: "Clear active filter",
  noFilteredChildren: "No child nodes match current filter",
  resultLimitExceeded: "There are more items than allowed limit of {{limit}}.",
  resultLimitExceededWithFiltering: "Please provide <link>additional filtering</link> - there are more items than allowed limit of {{limit}}.",
  increaseHierarchyLimit: "<link>Increase the hierarchy level size limit to {{limit}}.</link>",
  increaseHierarchyLimitWithFiltering: "Or, <link>increase the hierarchy level size limit to {{limit}}.</link>",
  unspecified: "Unspecified",

>   other: "Other",

Reviewer 2: These two used by `useTree` should be separated from the other ones. `useTree` can be used without `TreeRenderer` so in that case it does not make sense to require all these additional strings that are used only by renderer.

Author: They are separate, I just merged the object into one in the `README`. Split into two to be more clear.

Reviewer: I personally liked the non-split version better. Maybe adding a couple of comments would've been a better change?
```ts
const localizedStrings = {
  // strings for the `useTree` hook
  ...

  // strings for `TreeRenderer` and `TreeNodeRenderer`
  ...
};
```

Author: Reverted the change and left comments.

};

function MyTreeComponent({ imodelAccess, imodelKey }: { imodelAccess: IModelAccess; imodelKey: string }) {
  const { rootNodes, ...state } = useUnifiedSelectionTree({
    sourceName: "MyTreeComponent",
    imodelKey,
    imodelAccess,
    localizedStrings,
    getHierarchyDefinition: () => ({
      defineHierarchyLevel: async () => [],
    }),
  });
  if (!rootNodes) {

>     return "Loading...";

Reviewer: ```suggestion
    return localizedStrings.loading;
```

  }
  return <TreeRenderer rootNodes={rootNodes} localizedStrings={localizedStrings} {...state} />;
}
```

> To enable localization for the `TreeNodeRenderer` component, the localized strings should be supplied through `LocalizationContextProvider`:

Reviewer 2: ```suggestion
To enable localization for the custom `TreeNodeRenderer` component, the localized strings should be supplied through `LocalizationContextProvider`:
```

> To enable localization for the `TreeNodeRenderer` component, the localized strings should be supplied through `LocalizationContextProvider`:

Reviewer: Not sure this is correct. It's not the `TreeNodeRenderer` that has to be custom for the requirement to use `LocalizationContextProvider`, but the `TreeRenderer`. I suggest this:

```suggestion
In case the `TreeNodeRenderer` component is used within a custom tree renderer, the tree component should supply localized strings through `LocalizationContextProvider`:
```

Author: Updated based on suggestion.

```tsx

> const localizedStrings = {
>   loading: "Loading...",
>   filterHierarchyLevel: "Apply filter",
>   clearHierarchyLevelFilter: "Clear active filter",
>   noFilteredChildren: "No child nodes match current filter",
>   resultLimitExceeded: "There are more items than allowed limit of {{limit}}.",
>   resultLimitExceededWithFiltering: "Please provide <link>additional filtering</link> - there are more items than allowed limit of {{limit}}.",
>   increaseHierarchyLimit: "<link>Increase the hierarchy level size limit to {{limit}}.</link>",
>   increaseHierarchyLimitWithFiltering: "Or, <link>increase the hierarchy level size limit to {{limit}}.</link>",
> };
> 

Reviewer: The above example already shows what `localizedStrings` looks like, so I think we can omit it here for the sake of simplicity.

```suggestion
```

> export function TreeRenderer({ ...props }: TreeRendererProps) {

Reviewer: Let's rename to make it clear it's something that consumers would be creating.

```suggestion
export function MyTreeRenderer(props: MyTreeRendererProps) {
```
